Add support for username regex.

nbianca opened the pull request:
Hello,

I have added support for [Reserved username wildcard](https://meta.discourse.org/t/reserved-usernames-wildcard/41968/12).

First of all, I have extended the support to more than just a wildcard ~~, but to support any regex~~.

~~However, there is an issue with the username suggester. For example, if `i42-.*` is one of those restrictitions and I try to create a user called `i42-test`, the username suggester will try to suggest `i42-test1`, `i42-test2`, `i42-test3`, etc. and will not stop. I added another condition to stop suggesting after 10 attempts, however this is just a cheapfix. I would like to know how would the team want to fix this issue.~~

I look forward to hearing from you.

Thanks,
Bianca Nenciu

SamSaffron commented:
This changes semantics though

for example `sam.saffron` will now ban `sam9saffron` as well. 

Additionally, 

- We need some tests here to confirm behavior
- The if careful someone can add a site setting that will cause the code to explode, eg: `sam)`

SamSaffron commented:
I think the best way forward is simply to expand semantics here without full blown regex. 

Just add support for `*` explosion. So `*admin*` would be used to ban everyone with the word admin. This mean it will be a lot harder to get this wrong.

SamSaffron commented:
>  the username suggester will try to suggest i42-test1, i42-test2, i42-test3, etc. and will not stop.  

Yeah this is a security issue, it must be also addressed with the PR, if a * test is the one that fails then use a generic username maybe something like 

- first try usernameRANDOMNUMBER +1 +2 etc ... up to 100 times 
- then just give user random letters and try up to 100 times

at 200 bail out 

otherwise people will be able to crash any discourse instance with this * expansion enabled

nbianca commented:
Hello Sam,

I made the changes per your review:
* Regex is no longer supported. The only "special character" is the wildcard symbol. I escape any symbol that might be part of a regex, so `foo.bar` will not restrict `foo9bar`.
* I added tests for username availability.
* I also added generation of random usernames consisting of random characters after 100 fails. (I am generating a new username of the same length as the initial one.) Do you think I should generate a random username only if a wildcard restriction is applied? Also, I was thinking of creating random usernames like `adjective.noun.random_number` but I am not sure how this would be affected by localization.

SamSaffron commented:
```
attempt = rand(36 ** name.length).to_s(36)
```

looks somewhat odd to me... 

Perhaps simply:

```
attempt = SecureRandom.hex
```

SamSaffron commented:
`[0..name.length]` needs better bounds, cause if you pass a name that is 1000 chars long that will simply fail 1000 times. I would keep in simple and set length fixed to SiteSetting.max_username_length

SamSaffron commented:
I think all we have left is to use SiteSetting.max_username_length for username fallback length, except for that we should be all good

nbianca commented:
Sorry, I have been busy with university. I just pushed my changes. Tests should pass, so this should be ready to merge indeed.

SamSaffron commented:
@nbianca no worries at all an no rush! thanks so much for contributing to Discourse.

tgxworld commented on lib/user_name_suggester.rb:
```diff
@@ -21,12 +21,16 @@ def self.find_available_username_based_on(name, allow_username = nil)
       attempt = "#{name[0..max_length]}#{suffix}"
       i += 1
     end
+    until attempt == allow_username || User.username_available?(attempt) || i > 200
```
Based on the discussion in 

https://github.com/discourse/discourse/pull/4728#issuecomment-282862287. I think this should be `i > 100`?

nbianca replied:
The counter `i` is not reset before this so it is going to be 100 from the previous block and another 100 from this block.

tgxworld commented on spec/models/user_spec.rb:
```diff
@@ -475,6 +475,22 @@ def assert_good(username)
       expect(User.username_available?('test')).to eq(false)
     end
+
+    it 'tests when a username is reserved by an expression with wildcard' do
```
The description here doesn't sound right to me. Usually the `it` block describes what the expected result should be. In this case, I would go with something like:

```
context 'when reserved usernames contains an expression" do
  it "should return the right value when username is matched against the expresssion" do
  end
end
```

nbianca replied:
You are right. I changed the description.

tgxworld commented on spec/models/user_spec.rb:
```diff
@@ -475,6 +475,22 @@ def assert_good(username)
     end
+
+    it 'tests when a username is reserved by an expression with wildcard' do
+      SiteSetting.reserved_usernames = 'test)|*admin*|foo*|*bar|abc.def'
```
We'll need to update the description for the site settings to clearly indicate that Regexp expressions are allowed.

nbianca replied:
This is also correct. I have updated that [here](https://github.com/discourse/discourse/pull/4728/files#diff-aeb8651199334c1cb831e884ff85773aR1043).

tgxworld commented:
Thanks @nbianca
